File: daily/2024/2924_Find_Champion_II.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findChampion(int n, vector<vector<int>>& edges) {
        vector<int> indegree(n, 0);
        for (auto& i : edges) indegree[i[1]]++;
        int ans = -1;
        for (int i = 0; i < n; i++){
            if (indegree[i] == 0){
                if (ans == -1) ans = i;
                else return -1;
            }
        }
        return ans;
    }
};

int main(){
    int n = 3;
    vector<vector<int>> edges = {{0,1},{1,2}};
    // int n = 4;
    // vector<vector<int>> edges = {{0,2},{1,3},{1,2}};
    Solution S;
    cout << S.findChampion(n, edges) << endl;
    return 0;
}

File: daily_python/2924_Find_Champion_II.py
import os
import sys
from typing import List

class Solution:
    def findChampion(self, n: int, edges: List[List[int]]) -> int:
        indegree = [0] * n
        for edge in edges:
            indegree[edge[1]] += 1
        ans = -1
        for i in range(n):
            if indegree[i] == 0:
                if ans == -1:
                    ans = i
                else:
                    return -1
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            edges = [list(map(int, edge.split(','))) for edge in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((n, edges))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, edges in dataset:
        results.append(solution.findChampion(n, edges))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')